docs/0001-teleport-job-worker.md

```diff
@@ -0,0 +1,127 @@
+---
+authors: Brodie Kurczynski (brodie@kurczynski.org)
+state: draft
+---
+
+# RFD 1 - Teleport Job Worker
+
+## What
+
+A service that runs commands submitted by multiple users. These commands may run concurrently and will provide the users
+with the ability to limit resource (i.e. CPU, memory, and disk IO) usage. The service will consist of three pieces:
+
+* A reusable library implementing the functionality for working with jobs
+* An API server that wraps the functionality of the library
+* A command line interface (CLI) that communicates with the API server
+
+### Required Approvers
+
+* Engineering: (@sclevine && @gavinfrazar) || (@sclevine && @bernardjkim) || (@sclevine && @jimbishopp) || (
+  @gavinfrazar && @bernardjkim) || (@gavinfrazar @jimbishopp) || (@bernardjkim @jimbishopp)
+
+### Security
+
+Requests made to the service will use mTLS for authentication. Specifically:
+
+* TLS using the modern version 1.3
+* The cipher suite will need no configuration, it will be handled by `crypto/tls` (this was updated
+  in [Go 1.17](https://go.dev/doc/go1.17#crypto/tls))
+* X.509 public/private key pairs will be generated using RSA-4096
+
+For authorization, the service will use simple tokens to ensure that jobs can only be accessed and modified by the user
+who created them. This will be done by adding `grpc.UnaryInterceptor` to the API server to check if the user has
+authorization to interact with the resource. For the purposes of this challenge, this type of authorization will be
+fine. However, proper OAuth would need to be implemented for use in production.
+
+### UX
+
+The CLI used in the service will be a simple application that implements the shared library and expose all the
+functionality of the [API](#api).
+
+For example, calling `Query()` in the API is equivalent to calling:
+
+```shell
+job-worker query some-user-token
+```
+
+An exhaustive list of commands and arguments must be available using the `-help` flag.
+
+### Process Execution
+
+Jobs will be forked from the service's process and contain exactly one process that it manages. A job must have one of
+the statuses defined in `Statuses` of the [proto spec](../api/proto/job/job.proto) where it will start in an initial
+state and finish in one of three terminal states:
+
+```
+            ┌──► stopped  
+            │             
+running ────┼──► failed   
+            │             
+            └──► succeeded
+```
+
+Resource control will be implemented using cgroups. A parent directory `job-worker` will be created
+under `/sys/fs/cgroup` that will contain a directory for every job created. For example, the path for `job-a` would look
+like `/sys/fs/cgroup/job-worker/job-a`. Under this path, the following files would be updated to manage the resources
+for the job:
+
+* `cgroup.procs` contains the PID of the job's process being managed
```

Comment:
How will you ensure the cgroup applies to this PID for the entirety of its execution?

Reply:
I'm not sure I understand the question. The job will fork from the service, the job will apply the cgroups, and then execute the process given by the user.

The kernel [docs](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#processes) say:
> When a process forks a child process, the new process is born into the cgroup that the forking process belongs to at the time of the operation. After exit, a process stays associated with the cgroup that it belonged to at the time of exit until it’s reaped

So it should be applied the whole time.